File: src/dataset/Java/bank-account_exercism-uzilan/gemini/GeminiFlash_BankAccount_v2.java
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class BankAccount {

    private static final String WITHDRAWING_TOO_MUCH = "Cannot withdraw more money than is currently in the account";
    private static final String EMPTY_ACCOUNT = "Cannot withdraw money from an empty account";
    private static final String NEGATIVE_AMOUNT = "Cannot deposit or withdraw negative amount";
    private static final String ACCOUNT_CLOSED = "Account closed";
    private int account;
    private boolean open;
    private final Lock lock = new ReentrantLock();

    public void open() {
        lock.lock();
        try {
            this.open = true;
        } finally {
            lock.unlock();
        }
    }

    public void close() {
        lock.lock();
        try {
            this.open = false;
        } finally {
            lock.unlock();
        }
    }

    public int getBalance() throws BankAccountActionInvalidException {
        lock.lock();
        try {
            checkIfAccountIsClosed(open);
            return this.account;
        } finally {
            lock.unlock();
        }
    }

    public void deposit(int amount) throws BankAccountActionInvalidException {
        lock.lock();
        try {
            checkIfAccountIsClosed(open);
            checkForNegativeAmount(amount);
            this.account += amount;
        } finally {
            lock.unlock();
        }
    }

    public void withdraw(int amount) throws BankAccountActionInvalidException {
        lock.lock();
        try {
            checkIfAccountIsClosed(open);
            checkForNegativeAmount(amount);
            checkIfAccountIsZero(account);
            checkIfWithdrawingTooMuch(account, amount);
            this.account -= amount;
        } finally {
            lock.unlock();
        }
    }

    private void checkIfWithdrawingTooMuch(int account, int amount)
            throws BankAccountActionInvalidException {
        if (account - amount < 0) {
            throw new BankAccountActionInvalidException(WITHDRAWING_TOO_MUCH);
        }
    }

    private void checkIfAccountIsZero(int account)
            throws BankAccountActionInvalidException {
        if (account == 0) {
            throw new BankAccountActionInvalidException(EMPTY_ACCOUNT);
        }
    }

    private void checkForNegativeAmount(int amount)
            throws BankAccountActionInvalidException {
        if (amount < 0) {
            throw new BankAccountActionInvalidException(NEGATIVE_AMOUNT);
        }
    }

    private void checkIfAccountIsClosed(boolean open)
            throws BankAccountActionInvalidException {
        if (!open) {
            throw new BankAccountActionInvalidException(ACCOUNT_CLOSED);
        }
    }
}

File: src/dataset/python/bank-account_exercism-fruit-in/gemini/GeminiFlash_bank_account_v2.py
import threading


class BankAccount:
    def __init__(self):
        self._is_open = False
        self._balance = 0
        self._lock = threading.Lock()

    def get_balance(self):
        with self._lock:
            if not self._is_open:
                raise ValueError("Account not open")
            return self._balance

    def open(self):
        with self._lock:
            if self._is_open:
                raise ValueError("Account already open")
            self._is_open = True
            self._balance = 0

    def deposit(self, amount):
        with self._lock:
            if not self._is_open:
                raise ValueError("Account not open")
            if amount <= 0:
                raise ValueError("Cannot deposit non-positive amount")
            self._balance += amount

    def withdraw(self, amount):
        with self._lock:
            if not self._is_open:
                raise ValueError("Account not open")
            if amount <= 0:
                raise ValueError("Cannot withdraw non-positive amount")
            if amount > self._balance:
                raise ValueError("Insufficient funds")
            self._balance -= amount

    def close(self):
        with self._lock:
            if not self._is_open:
                raise ValueError("Account not open")
            self._is_open = False